#16503: Optimize CoreRangeSets for CBs and semaphores

jbaumanTT opened the pull request:
### Ticket
#16503 

### Problem description
CoreRangeSets passed in from users may contain more CoreRanges than needed to properly represent the range, which can lead to inefficiency and redundant multicasts. 

### What's changed
Optimize the CoreRangeSets used for CBs and semaphores to improve that.

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) tests passes
- [ ] New/Existing tests provide coverage for changes

jbaumanTT committed:
#16503: Optimize CoreRangeSets for CBs and semaphores

CoreRangeSets passed in from users may contain more CoreRanges than needed to
properly represent the range, which can lead to inefficiency and redundant
multicasts. Optimize the CoreRangeSets used for CBs and semaphores to improve
that.

tt-asaigal reviewed (approved):
I understand the overall intention and why this helps. I wonder if calling the API `optimize` is misleading. Perhaps we should use something like `merge_with_self()` or something along those lines.
Also I think we should add a comment wherever this API is used (and in the implementation), briefly mentioning how this helps program dispatch. I can see developers getting confused later on.

jbaumanTT commented:
> I understand the overall intention and why this helps. I wonder if calling the API `optimize` is misleading. Perhaps we should use something like `merge_with_self()` or something along those lines. Also I think we should add a comment wherever this API is used (and in the implementation), briefly mentioning how this helps program dispatch. I can see developers getting confused later on.

Ok, I've renamed it to `merge_ranges()` to be clearer about it's supposed to do, and added more comments everywhere.
